File: assignment2/q1_2.py
import matplotlib.pyplot as plt

# Define signal points
s1 = (1, 0)
s2 = (1, (1/3)**0.5)
s3 = (-2, 0)
s4 = (-2, (1/3)**0.5)

# Calculate axis limits
x_min, x_max = min(s1[0], s2[0], s3[0], s4[0]), max(s1[0], s2[0], s3[0], s4[0])
y_min, y_max = min(s1[1], s2[1], s3[1], s4[1]), max(s1[1], s2[1], s3[1], s4[1])

# Determine center of axes
x_center = (x_min + x_max) / 2
y_center = (y_min + y_max) / 2

# Plot signal constellation points
plt.figure(figsize=(8, 8))
plt.scatter(*s1, color='red', label='s1 (1, 0)')
plt.scatter(*s2, color='blue', label='s2 (1, √(1/3))')
plt.scatter(*s3, color='green', label='s3 (-2, 0)')
plt.scatter(*s4, color='purple', label='s4 (-2, √(1/3))')

# Annotate points
plt.text(s1[0], s1[1] + 0.1, 's1', color='red', fontsize=12)
plt.text(s2[0], s2[1] + 0.1, 's2', color='blue', fontsize=12)
plt.text(s3[0], s3[1] + 0.1, 's3', color='green', fontsize=12)
plt.text(s4[0], s4[1] + 0.1, 's4', color='purple', fontsize=12)

# Define and annotate decision regions
plt.text((x_center + x_max) / 2, (y_center + y_max) / 2, 'I1', color='black', fontsize=14)
plt.text((x_center + x_min) / 2, (y_center + y_max) / 2, 'I2', color='black', fontsize=14)
plt.text((x_center + x_min) / 2, (y_center + y_min) / 2, 'I3', color='black', fontsize=14)
plt.text((x_center + x_max) / 2, (y_center + y_min) / 2, 'I4', color='black', fontsize=14)

# Add gridlines and centered axes
plt.axhline(y=y_center, color='black', linewidth=0.5, linestyle='--')
plt.axvline(x=x_center, color='black', linewidth=0.5, linestyle='--')

# Formatting the plot
plt.xlim(x_min - 1, x_max + 1)
plt.ylim(y_min - 0.5, y_max + 0.5)
plt.xticks([])  # Hide x ticks
plt.yticks([])  # Hide y ticks
plt.grid(True, linestyle='--', alpha=0.7)
plt.title("Signal Constellation Diagram", fontsize=14)
plt.xlabel("φ1 Axis", fontsize=12)
plt.ylabel("φ2 Axis", fontsize=12)
plt.legend()
plt.gca().set_aspect('equal', adjustable='box')

# Show the plot
plt.show()

File: assignment2/q1_2.m
% MATLAB Script to Plot Signal Constellation Diagram using polarplot and polarscatter

% Define energy of the signals (E1 = 1 for simplicity)
E1 = 1;

% Define signal vectors (radii and angles in polar coordinates)
radii = [sqrt(E1), sqrt(E1), 2*sqrt(E1), 2*sqrt(E1)]; % Magnitudes
angles = [0, atan(sqrt(1/3)), pi, pi + atan(sqrt(1/3))]; % Angles in radians

% Convert to Cartesian coordinates for additional plotting
[x, y] = pol2cart(angles, radii);

% Plotting the signal constellation using polarscatter
figure;
h = polaraxes;
h.ThetaZeroLocation = 'top';
h.ThetaDir = 'counterclockwise';

% Scatter points
polarscatter(angles, radii, 100, 'filled', 'r');
hold on;

% Label points with their respective signal vectors
text(x(1), y(1), 's_1', 'VerticalAlignment', 'bottom', 'HorizontalAlignment', 'left');
text(x(2), y(2), 's_2', 'VerticalAlignment', 'bottom', 'HorizontalAlignment', 'left');
text(x(3), y(3), 's_3', 'VerticalAlignment', 'top', 'HorizontalAlignment', 'right');
text(x(4), y(4), 's_4', 'VerticalAlignment', 'top', 'HorizontalAlignment', 'right');

% Add grid for better clarity
polarplot([0, angles(1)], [0, radii(1)], 'k--');
polarplot([0, angles(2)], [0, radii(2)], 'k--');
polarplot([0, angles(3)], [0, radii(3)], 'k--');
polarplot([0, angles(4)], [0, radii(4)], 'k--');

% Overlay Cartesian axes for clarity
cartesian_axes_limits = 2.5; % Define limits for the Cartesian axes
plot([-cartesian_axes_limits, cartesian_axes_limits], [0, 0], 'k-', 'LineWidth', 1.5); % X-axis
plot([0, 0], [-cartesian_axes_limits, cartesian_axes_limits], 'k-', 'LineWidth', 1.5); % Y-axis

% Label axes
txt_offset = 0.2; % Offset for axis labels
text(cartesian_axes_limits - txt_offset, 0, 'x', 'FontSize', 10, 'HorizontalAlignment', 'center');
text(0, cartesian_axes_limits - txt_offset, 'y', 'FontSize', 10, 'HorizontalAlignment', 'center');

% Add decision regions (triangular boundaries for simplicity)
fill([x(1), x(2), 0], [y(1), y(2), 0], 'b', 'FaceAlpha', 0.1, 'EdgeColor', 'none'); % Region for s1
fill([x(2), x(4), 0], [y(2), y(4), 0], 'g', 'FaceAlpha', 0.1, 'EdgeColor', 'none'); % Region for s2
fill([x(3), x(1), 0], [y(3), y(1), 0], 'r', 'FaceAlpha', 0.1, 'EdgeColor', 'none'); % Region for s3
fill([x(4), x(3), 0], [y(4), y(3), 0], 'y', 'FaceAlpha', 0.1, 'EdgeColor', 'none'); % Region for s4

% Add title
title('Signal Constellation Diagram with Decision Regions');

% Customize and hold off
hold off;
